src/linker/Linker/LinkContext.cs

```diff
@@ -502,10 +502,14 @@ public void LogMessage (MessageContainer message)
 				message.Category == MessageCategory.Info) && !LogMessages)
 				return;
 
-			if (message.Category == MessageCategory.Warning &&
-				NoWarn.Contains ((uint) message.Code)) {
-				// This warning was turned off by --nowarn.
-				return;
+			if (message.Category == MessageCategory.Warning) {
+				if (NoWarn.Contains ((uint) message.Code))
+					// This warning was turned off by --nowarn.
+					return;
+
+				if ((GeneralWarnAsError && (!WarnAsError.TryGetValue ((uint) message.Code, out var warnAsError) || warnAsError)) ||
+					(!GeneralWarnAsError && (WarnAsError.TryGetValue ((uint) message.Code, out warnAsError) && warnAsError)))
+					message = MessageContainer.CreateErrorMessage (message.Text, message.Code.Value, message.SubCategory, message.Origin, isWarnAsError: true, message.Version);
```

Comment:
It would be nice to factor this so that the check for the code range (in Create*Message) happens just once when the warning/error is created, and not have to track isWarnAsError/version on the error.

Reply:
I don't want to block the PR on it though. Also related to https://github.com/mono/linker/issues/1384.

Reply:
I agree we should factor this. This is also related to #1406.